src/core/reactor.cc

```diff
@@ -2680,9 +2683,19 @@ class reactor::io_queue_submission_pollfn final : public reactor::pollfn {
     // the future
     timer<> _nearest_wakeup { [this] { _armed = false; } };
     bool _armed = false;
+    steady_clock_type::time_point _last_poll;
 public:
-    io_queue_submission_pollfn(reactor& r) : _r(r) {}
+    io_queue_submission_pollfn(reactor& r)
+        : _r(r)
+        , _last_poll(steady_clock_type::now())
+    {}
     virtual bool poll() final override {
+        auto now = steady_clock_type::now();
+        auto delta = now - _last_poll;
+        _last_poll = now;
+        if (delta > _r._io_latency_goal) {
+            _r._io_latency_goal_violation_time += delta - _r._io_latency_goal;
```

Comment:
I don't think this is accurate. If we poll at exactly latency_goal intervals, the disk would be underloaded. We stuff the disk with enough requests for calculated io_latency_goal, but some requests can't be executed in parallel, and we'll io-stall on the time it takes for delivery of the last responses.

What we really need is a histogram of io poll times, but we know the problem with that.

Reply:
> If we poll at exactly latency_goal intervals, the disk would be underloaded

Scheduler puts 1.5 times more requests into disk than it expects to complete after latency goal, so it won't be underloaded.

> We stuff the disk with enough requests for calculated io_latency_goal, but some requests can't be executed in parallel, and we'll io-stall on the time it takes for delivery of the last responses.

Why stall? Scheduler puts requests in batches, but frees the resource as requests appear completed and puts more data back into disk as the resource gets released.

> What we really need is a histogram of io poll times, but we know the problem with that.

Did recent Amnon's improvements over historgrams made enough progress to reconsider that?
Another option -- runtime switch to report (or not) the histograms. Like on-demand tracing on/off, I don't think we'll lose much information if turn on CPU poll rates and IO poll rates histograms for just 5 minutes.
Collecting shouldn't be a big deal. Or is it?

Reply:
@amnonh ?

Reply:
If the execution time of a request is (request_cost / disk_capacity_per_sec)  (as defined by our model), then this is accurate. But in practice, the execution time is (request_cost / disk_capacity_per_sec) * some_random_variable, which  has its probability distribution centered around 1, but also has outliers >> 1.

If the tail of the probability distribution has low overall probability, then I don't think it has much effect, but if it's high it may stall the queue. For such disks we need higher latency goal.

Reply:
Here's the probability distribution from some random disk:

```
@usecs[271581194, R]: 
[0]                    1 |                                                    |
[1]                    6 |                                                    |
[2, 4)                16 |                                                    |
[4, 8)                27 |                                                    |
[8, 16)               56 |                                                    |
[16, 32)             104 |                                                    |
[32, 64)             190 |                                                    |
[64, 128)            671 |@@                                                  |
[128, 256)         12892 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[256, 512)          3861 |@@@@@@@@@@@@@@@                                     |
[512, 1K)            123 |                                                    |
[1K, 2K)              11 |                                                    |
[2K, 4K)               1 |                                                    |
```

Requests that take more than the io latency goal will starve the disk for some time. But they're pretty rare.

Reply:
ping @amnonh see question above

Reply:
We have multiple ways of handling a distribution that are already in place and a few more that are cumming.
1. Summaries, it would give some data, but I think the quantile will not help and the distribution in this case is more import.
2. aggregating histograms, will a single histogram per node would suffice? In Scylladb/PR#13293 based on the configuration, Scylla expose aggregate or non aggregate metrics.
3. Use low resolution histogram. In @avikivity example,  a histogram with 2 coeficiant where less than 10 buckets were used was enough, we can use a histogram with a minimal number of buckets that would still expose enough of the distribution.

Future enhencement would add enable and disable metrics in run time and even furthere down the line are better histogram support but that would take longer.